python-client/tests/test_performance.py

```diff
@@ -236,6 +236,25 @@ def test_diff_accuracy(data, model, threshold, expected_metric, request):
     assert result.passed
 
 
+@pytest.mark.parametrize(
+    "data,model,threshold,expected_metric",
+    [("german_credit_data", "german_credit_always_default_model", 0, 0)],
+)
+def test_diff_accuracy_fail(data, model, threshold, expected_metric, request):
+    tests = GiskardTestFunctions()
+    data = request.getfixturevalue(data)
+    result = tests.performance.test_diff_accuracy(
+        actual_slice=data.slice(lambda df: df[df.sex == "male"]),
+        reference_slice=data.slice(lambda df: df[(df.sex == "female") & (df.default == "Not default")]),
+        model=request.getfixturevalue(model),
+        threshold=threshold,
+    )
+    assert round(result.metric, 2) == expected_metric
+    assert type(result.output_df) is bytes
+    assert not result.passed
+    assert result.messages
```

Comment:
being precise in the assertions helps with increasing test sensitivity. Here we'd need to assert the actual message too (and maybe the size of the messages array).
you can use `re` module to assert the message content with `re.match`

Also could you test the rest of the `test_diff_*` functions in the same way? Since they have the same signature, you can run tests in a loop like:

```
    for test_fn in [
        tests.performance.test_diff_accuracy,
        tests.performance.test_diff_f1,
        tests.performance.test_diff_precision,
        tests.performance.test_diff_recall,
        tests.performance.test_diff_reference_actual_f1
    ] :
        result = test_fn(
            actual_slice=data.slice(lambda df: df[df.sex == "male"]),
            reference_slice=data.slice(lambda df: df[(df.sex == "female") & (df.default == "Not default")]),
            model=request.getfixturevalue(model),
            threshold=threshold,
        )
     ...   
        
```

or maybe even parameterize a test name, but you'd have to then convert a string with a function name to an actual test function before running it